switch to boto3 and add support for application ELBs with target groups.

jonhadfield opened the pull request:
ISSUE TYPE

Feature Pull Request
COMPONENT NAME

ec2_asg

ANSIBLE VERSION

ansible 2.2.0.0 (detached HEAD cdec853e37) last updated 2016/11/01 12:32:01 (GMT +100)
  lib/ansible/modules/core: (detached HEAD fe9c56a003) last updated 2016/11/01 12:32:05 (GMT +100)
  lib/ansible/modules/extras: (detached HEAD f564e77a08) last updated 2016/11/01 12:32:05 (GMT +100)
  config file =
  configured module search path = Default w/o overrides
SUMMARY

Upgrade module to use boto3 in order to support new functionality.
Added support for Application ELBs via Target Groups
Compliments the following PRs in that it enables creation of an ASG with ansible generated application load balancers and target groups:
elb_target_group: ansible/ansible-modules-extras#2947
elb_application_lb: ansible/ansible-modules-extras#2946

Note:

Only significant logic change is to support the integration with target groups/application lbs, so any existing issues remain
Only minimal testing performed

jonhadfield committed:
switch to boto3 and add support for application ELBs with target groups.

jonhadfield commented:
This is a replacement for https://github.com/ansible/ansible-modules-core/pull/5746 after failed attempts at moving the original using PR mover.

jonhadfield committed:
use py23 compatible dict iterator.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -441,173 +510,272 @@ def create_autoscaling_group(connection, module):
-                 health_check_type=health_check_type,
-                 default_cooldown=default_cooldown,
-                 termination_policies=termination_policies)
+        launch_configs = connection.describe_launch_configurations(LaunchConfigurationNames=[launch_config_name])
```
Needs error handling here for if no launch configs match the search, since later the `[0]` element is used without bounds-checking.

s-hertel replied:
Fixed this on the line below.
describe_launch_configurations() is returning something like 
```
{'LaunchConfigurations': [], 'ResponseMetadata': {'RetryAttempts': 0, 'HTTPStatusCode': 200, 'RequestId': '57dccc7c-2080-11e7-8fbd-adf472434545', 'HTTPHeaders': {'x-amzn-requestid': '57dccc7c-2080-11e7-8fbd-adf472434545', 'date': 'Thu, 13 Apr 2017 19:35:25 GMT', 'content-length': '350', 'content-type': 'text/xml'}}}"
```
So it always had longer than a length of 0 and the below check had no effect. Added a check for the same thing on lines 719-720 too.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -441,173 +510,272 @@ def create_autoscaling_group(connection, module):
-            suspend_processes(ag, module)
+            connection.create_auto_scaling_group(**ag)
+            ag = connection.describe_auto_scaling_groups(
+                AutoScalingGroupNames=[group_name])['AutoScalingGroups'][0]
```
This also needs bounds checking and (maybe) retries if the ASG isn't yet available.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -441,173 +510,272 @@ def create_autoscaling_group(connection, module):
+                        notification_types,
+                    ]
+                )
+            as_group = connection.describe_auto_scaling_groups(
```
It looks like this call and the above (line 570) call are filling 2 variables with the same information. Is there a reason two names are used?

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -441,173 +510,272 @@ def create_autoscaling_group(connection, module):
-            if have_tags != want_tags:
+                    dead_tags.append(dict(ResourceId=as_group['AutoScalingGroupName'],
+                                          ResourceType='auto-scaling-group', Key=have_tag_keyval))
+                    have_tags = [have_tag for have_tag in have_tags if have_tag['Key'] != have_tag_keyval]
```
This is a bit confusing since it's assigning to the same name name as comes in. I think this section could be cleared up with the use of `set`. So something along the lines of:

```
for dead_tag in set(have_tag_keyvals).difference(want_tag_keyvals):
    dead_tags.append(..... and so on ...)
```

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -441,173 +510,272 @@ def create_autoscaling_group(connection, module):
+            # module.fail_json(msg=list(wanted_elbs))
+
+            has_elbs = set(as_group['LoadBalancerNames'])
+            # module.fail_json(msg=wanted_elbs)
```
Please remove these commented out fail_json calls.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -441,173 +510,272 @@ def create_autoscaling_group(connection, module):
+            # check if all requested are already existing
+            if has_elbs.issuperset(wanted_elbs):
+                # if wanted contains less than existing, then we need to delete some
+                elbs_to_detach = has_elbs.difference(wanted_elbs)
```
Yay sets!

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -896,18 +1071,23 @@ def main():
-    if not HAS_BOTO:
-        module.fail_json(msg='boto required for this module')
+    if not HAS_BOTO_3:
+        module.fail_json(msg='boto3 required for this module')
```
:tada: boto3!

s-hertel committed:
removing commented out fail_json calls

utilize sets to simplify logic

remove setting a redundant variable

add bounds checking in two places

add AWSRetry decorator - do we want this for other functions too?

change xrange to range so python3 doesn't fail

remove sorting lists of dicts; in python2 this returns None, in python3 this fails

s-hertel committed:
remove error variable from traceback.format_exc

ryansb committed:
Remove boto2-style calls brought in by rebase

Old boto-style calls to `as_group` attributes break in boto3

Also remove module from legacy-PEP8 list

ryansb committed:
Add  parameter to target_group_arn option

willthames commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -443,146 +517,243 @@ def create_autoscaling_group(connection, module):
     asg_tags = []
     for tag in set_tags:
-        for k,v in tag.items():
+        for k,v in iteritems(tag):
```
Not sure I understand this change

s-hertel replied:
I wondered about this too. I think he chose this because iterating over the key/values in a dict has changed in python3 and module_utils.six checks for the python version so the behavior is consistent. It seems to behave appropriately. https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/six.py#L367
Do you think this might end up being a problem?

jonhadfield replied:
@s-hertel Exactly that. Thought it best to make use of six for future compatibility with python3.

ryansb replied:
In Ansible we typically just use `.items()` because as a loop iterator there's no (usage) difference between the Py2 list and Py3 iterable behavior, and if your dictionary is less than 1000 items or so, the iterable behaviour is actually *slower* in Python 2 (trading time for space, naturally). Can you please leave it as `items` instead?

willthames commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -902,18 +1080,23 @@ def main():
 
-    if not HAS_BOTO:
-        module.fail_json(msg='boto required for this module')
+    if not HAS_BOTO_3:
```
should be HAS_BOTO3, which you can import from module_utils.ec2

s-hertel replied:
Nice, will fix.

willthames commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -443,146 +517,243 @@ def create_autoscaling_group(connection, module):
     default_cooldown = module.params.get('default_cooldown')
     wait_for_instances = module.params.get('wait_for_instances')
-    as_groups = connection.get_all_groups(names=[group_name])
+    as_groups = connection.describe_auto_scaling_groups(AutoScalingGroupNames=[group_name])
```
All AWS calls should be wrapped in a `try`/`except` block

ryansb replied:
I'm fine with this being a followup PR. Error handling needs to be added, but that'd increase the diffstat of this PR pretty significantly, and since the handling wasn't there before I wouldn't block the merge over it.

ryansb replied:
Strictly, they should also be using AWSRetry as well, but that can be left to a followup PR.

jonhadfield replied:
I'm calling this 16 times in total:
`connection.describe_auto_scaling_groups(AutoScalingGroupNames=[group_name])['AutoScalingGroups'][0]`
Will move it out to a function that includes a try/except block and then take a look at the remaining boto3 calls.

jonhadfield replied:
@ryansb Just seen your message re: diffstat. Will hold off on refactoring.

s-hertel committed:
Fix HAS_BOTO3 check

s-hertel committed:
use tags.items() instead of iteritems

s-hertel committed:
import botocore

jonhadfield commented:
Anything you're waiting on me for?

s-hertel commented:
There are merge conflicts now to rebase, but not that I'm aware of. I think feature freeze was to blame. This still looks good to me.

vanga committed:
Fixed bugs in deleting autoscaling groups

vanga commented:
There are some bugs in delete_autoscaling_group function. 
I have fixed and created a pull request for @jonhadfield branch.

s-hertel committed:
make changes in deleting autoscaling groups pep8

s-hertel committed:
more pep8

s-hertel committed:
fix version

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -244,25 +248,23 @@
-log.getLogger('boto').setLevel(log.CRITICAL)
-#log.basicConfig(filename='/tmp/ansible_ec2_asg.log',level=log.DEBUG, format='%(asctime)s: %(message)s', datefmt='%m/%d/%Y %I:%M:%S %p')
-
+from ansible.module_utils.six import iteritems
```
This isn't used anywhere

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_asg.py:
```diff
@@ -276,150 +278,214 @@ def enforce_required_arguments(module):
     return properties
 
+
 def elb_dreg(asg_connection, module, group_name, instance_id):
```
What does this method name mean? No need to do anything, I'm just curious.

ryansb replied:
Oh, it's supposed to be "dereg" as in "deregister"

ryansb reviewed (commented):
Looks good - I'll have to do some testing, but all feedback is there as a note to myself on things to fix later.

tomislacker commented:
@jonhadfield I've also added metrics collection support in jonhadfield/ansible#2

tomislacker committed:
Fix mutually exclusive arguments
Only one of the AvailabilityZones and VPCZoneIdentifier arguments should be provided to the CreateAutoScalingGroup call.

tomislacker committed:
Fix notification setup

s-hertel commented:
@tomislacker
Thanks for this! Can you please move your metrics collection support to a new PR since it's adding a new feature? I believe I've fixed the bugs but you might want to check me on that. You could keep the Fix notification setup commit though, and I'll add that. Good catch. :)

s-hertel committed:
fix bugs so local integration tests run

s-hertel committed:
fix launch config check

jonhadfield committed:
reflect changed status for ASG updates

jonhadfield commented:
Just re-tested after all the changes and needed to add a fix https://github.com/ansible/ansible/pull/19667/commits/1aa044f5efe4f196f8e763b889992e01ade1eb46 as changes to an existing ASG weren't being reflected in 'changed' status.

@tomislacker I've tested your PR https://github.com/jonhadfield/ansible/pull/2 for adding metrics collection support (locally only, not merged) and looks good! Only initial comments are:
- metrics_collection in docs is missing the 'default' 
- assuming additional metrics cost, shouldn't this be disabled by default?

@s-hertel asked to move it to a separate PR, so let me know when done and I'll re-test.

s-hertel committed:
Fix existing exception handling and use traceback.

Fix imports

s-hertel committed:
line length

s-hertel commented:
I removed all traces of boto.exception.stuff since it's no longer going to work here. I also didn't replace the exception handling around boto3_conn since boto3 fails when we try to use the connection rather than when it is created (whereas boto2 would give a NoAuthHandlerFound error when unable to get the connection). I kept a check in main in case the profile provided doesn't exist at all or bad credentials are provided. We still need to wrap the rest of the AWS calls in a try/except block as discussed above in a follow up PR - I just fixed the existing ones. 
Since this is passing the ec2_asg integration tests this gets a shipit from me.

@ryansb @willthames see any blockers for this now?

ryansb commented:
shipit

ryansb commented:
shipit

tomislacker commented:
@jonhadfield I've updated my branch for jonhadfield/ansible#2 for your review. Sorry for the delays, been a little too busy lately. Once this PR is merged, I can open one up on this project instead.

shawnsi commented:
I've uncovered a potential concern with this update while integration testing.  In short, there are some parameters that worked implicitly and no longer do.  For example, `launch_config_name` would be pulled from any pre-existing ASG when the module was used.  This breaks plays like this:

```
      ec2_asg:
        name: "my-asg"
        region: "us-east-1"
        health_check_type: ELB
```

This type of behavior is something used in my scenarios to toggle notifications, health checking, etc. back on after deployment activities.  This is a strange edge case as the parameters are marked required in documentation but the module was functional without them anyhow.  I'm open to refactoring our usage but it will be painful.  I imagine other users could hit the same problems as well.

Here is the error I see related to this when testing the new module:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Invalid type for parameter LaunchConfigurationNames[0], value: None, type: <type 'NoneType'>, valid types: <type 'basestring'>fatal: [us-east-1-xxx]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_padsBj/ansible_module_ec2_asg.py\", line 1302, in <module>\n    main()\n  File \"/tmp/ansible_padsBj/ansible_module_ec2_asg.py\", line 1291, in main\n    create_changed, asg_properties = create_autoscaling_group(connection, module)\n  File \"/tmp/ansible_padsBj/ansible_modlib.zip/ansible/module_utils/cloud.py\", line 103, in retry_func\nbotocore.exceptions.ParamValidationError: Parameter validation failed:\nInvalid type for parameter LaunchConfigurationNames[0], value: None, type: <type 'NoneType'>, valid types: <type 'basestring'>\n", "module_stdout": "", "msg": "MODULE FAILURE"}```

s-hertel commented:
Hi @shawnsi. Wow, I didn't even know that was functionality we had. `launch_config_name` is specified as required. My local integration tests do not test that edge case. I'll see if that's able to be reproduced with boto3 and I'll make a fix (and correct the documentation) if so. Thanks for being another set of eyes on this.

@tomislacker No problem, thanks for doing that! Looks great. I will cherry pick fd68a1078a1841aad22f43ef3dfb0de3aa68fc6f and 9705587401878d4f9ba229ea2be739ce9d6bfbc2.

s-hertel committed:
Allow desired_capacity, min_size, max_size, launch_config_name to be derived from the existing ASG if not specified

Remove code updating dict after ASG already uses it

s-hertel commented:
@shawnsi Can you please run your tests again? I think it should work now.

tomislacker commented:
@s-hertel I was able to successfully test out the workflow that @shawnsi mentioned previously.

s-hertel commented:
shipit

@ryansb Can you check this out again when you get the chance?

ryansb commented:
shipit

avishefi commented:
Hi,
I noticed that ec2_asg doesn't deregister instances from ALB target groups.
Is this by design? Do ELB fail to deregister instances but ALBs do that out of the box and therefore this was not implemented in ec2_asg?

willthames commented:
@avishefi - I'd think deregistering instances should be done through elb_target_groups - you disassociate the ASG from the TG.

If there's a use case that isn't working, let us know, preferably in a separate issue now that this  one is closed.
